File: androidApp/src/main/java/com/wonddak/loacell/android/ui/home/BottomAppBar.kt
package com.wonddak.loacell.android.ui.home

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.BottomAppBar
import androidx.compose.material3.BottomAppBarDefaults
import androidx.compose.material3.FloatingActionButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.SmallFloatingActionButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import com.wonddak.loacell.SharedRes
import com.wonddak.loacell.android.ui.common.MyIconButton
import com.wonddak.loacell.android.ui.common.MyRoomIconButton
import com.wonddak.loacell.android.viewModel.LoaCellViewModel
import com.wonddak.loacell.ext.checkTimeOver
import com.wonddak.loacell.model.Dialog
import com.wonddak.loacell.model.RoomRole
import com.wonddak.loacell.model.RoomState
import com.wonddak.loacell.model.Sheet
import com.wonddak.loacell.store.CommonRaidHelper


@Composable
fun BottomAppBar(
    loaCellViewModel: LoaCellViewModel
) {
    val role = loaCellViewModel.myRole
    val totalRoomInfo = loaCellViewModel.totalRoomInfoValue
    val focusUserInfo = totalRoomInfo.userInfo
    val focusRaidInfo  = totalRoomInfo.raidInfo
    val tabState = totalRoomInfo.tabState

    val selectedRoomId = loaCellViewModel.roomId
    val syncData = loaCellViewModel.syncData

    loaCellViewModel.apply {
        BottomAppBar(
            floatingActionButton = {
                AnimatedVisibility(!(tabState == RoomState.Setting || showSetting)) {
                    SmallFloatingActionButton(
                        content = {
                            if (focusUserInfo != null || focusRaidInfo != null) {
                                Icon(
                                    painter = painterResource(id = SharedRes.images.delete.drawableResId),
                                    contentDescription = null,
                                    modifier = Modifier.size(18.dp)
                                )
                            } else {
                                Icon(Icons.Filled.Add, null)
                            }
                        },
                        onClick = {
                            bottomAddAction()
                        },
                        containerColor = BottomAppBarDefaults.bottomAppBarFabColor,
                        elevation = FloatingActionButtonDefaults.bottomAppBarFabElevation(2.dp),
                    )
                }
            },
            actions = {
                if (selectedRoomId.isEmpty()) {
                    AnimatedVisibility(!loaCellViewModel.showSetting) {
                        Row() {
                            MyIconButton(
                                imageResource = SharedRes.images.refresh,
                                enabled = !syncData
                            ) {
                                syncStart()
                            }
                        }
                    }
                } else {
                    AnimatedVisibility(
                        (focusUserInfo == null) && (focusRaidInfo == null),
                    ) {
                        Row() {
                            MyRoomIconButton(
                                loaCellViewModel = loaCellViewModel,
                                state = RoomState.Raid
                            )
                            MyRoomIconButton(
                                loaCellViewModel = loaCellViewModel,
                                state = RoomState.User
                            )
                            if (role == RoomRole.OWNER || role == RoomRole.MANAGER) {
                                MyRoomIconButton(
                                    loaCellViewModel = loaCellViewModel,
                                    state = RoomState.Setting
                                )
                            }
                        }

                    }
                    AnimatedVisibility(
                        focusUserInfo != null
                    ) {
                        focusUserInfo?.let {
                            Row() {
                                MyIconButton(
                                    imageResource = SharedRes.images.change_person
                                ) {
                                    showDialog(Dialog.CHARACTER_EDIT)
                                }
                                MyIconButton(
                                    imageResource = SharedRes.images.refresh,
                                    enabled = it.checkTimeOver()
                                ) {
                                    updateCharacter(selectedRoomId,it)
                                }
                            }
                        }
                    }
                    AnimatedVisibility(
                        focusRaidInfo != null
                    ) {
                        Row() {
                            focusRaidInfo?.let { info ->
                                MyIconButton(
                                    imageResource = SharedRes.images.room_setting
                                ) {
                                    showDialog(Sheet.RAID_EDIT)
                                }
                                val icon = if (info.isFinish) {
                                    SharedRes.images.task_finish_done
                                } else {
                                    SharedRes.images.task_finish_not
                                }
                                MyIconButton(
                                    imageResource = icon
                                ) {
                                    CommonRaidHelper.updateFinish(
                                        info.roomId,
                                        info.raidId,
                                        !info.isFinish
                                    )
                                }
                            }
                        }
                    }
                }
            },
        )
    }
}

File: iosApp/iosApp/ui/home/BottomAppBar.swift
//
//  BottomAppBar.swift
//  iosApp
//
//  Created by WonHee Jung on 2023/10/12.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

struct BottomAppBar: View {
    @EnvironmentObject var viewModel: LoaCellViewModel
    let height :CGFloat
    let iconSize : CGFloat = 30
    
    var totalRoomInfo : TotalRoomInfo {
        viewModel.totalRoomInfo
    }
    
    var tabState : RoomState {
        totalRoomInfo.tabState
    }
    var showSetting :Bool {
        viewModel.showSetting
    }

    
    var body: some View {
        HStack {
            HStack {
                if (viewModel.roomId.isEmpty) {
                    if (!showSetting) {
                        HStack {
                            IconButton(resource: \.refresh) {
                                viewModel.syncStart(force: false)
                            }
                        }
                    }
                }else {
                    if (viewModel.userInfo == nil && viewModel.raidInfo == nil) {
                        HStack {
                            IconButton(resource: \.room) {
                                viewModel.setTabStatus(state: RoomState.raid)
                            }
                            IconButton(resource: \.person) {
                                viewModel.setTabStatus(state: RoomState.user)
                            }
                            if (viewModel.myRole == RoomRole.owner || viewModel.myRole == RoomRole.manager) {
                                IconButton(resource: \.room_setting) {
                                    viewModel.setTabStatus(state: RoomState.setting)
                                }
                            }
                            
                        }
                    } else if(viewModel.userInfo != nil) {
                        HStack {
                            IconButton(resource: \.change_person) {
                                viewModel.showDialog(modal: shared.Dialog.characterEdit)
                            }
                            IconButton(resource: \.refresh, enabled: viewModel.userInfo!.checkTimeOver()) {
                                viewModel.updateCharacter(roomId: viewModel.roomId, userInfo: viewModel.userInfo!)
                            }.disabled(!viewModel.userInfo!.checkTimeOver())
                        }
                    } else if(viewModel.raidInfo != nil) {
                        HStack {
                            IconButton(resource: \.room_setting) {
                                viewModel.showDialog(modal: Sheet.raidEdit)
                            }
                            let info = viewModel.raidInfo!
                            IconButton(resource: info.isFinish ? \.task_finish_done : \.task_finish_not) {
                                CommonRaidHelper().updateFinish(
                                    roomId: info.roomId,
                                    raidId: info.raidId,
                                    isFinish: !info.isFinish
                                )
                            }
                        }
                    }
                }
            }
            Spacer()
            if (tabState == RoomState.setting || showSetting) {
                
            } else {
                Button {
                    viewModel.bottomAddAction()
                } label: {
                    if !viewModel.focusUserName.isEmpty || !viewModel.focusRaidId.isEmpty {
                        Image(resource: \.delete_)
                            .resizable()
                            .foregroundColor(.black)
                            .frame(width: 18, height: 18)
                    } else {
                        Image(systemName: "plus")
                            .foregroundColor(.black)
                    }
                    
                }
                .frame(width: 40, height: 40)
                .background(ColorManager.BackgroundContainerColor)
                .shadow(radius: 5)
                .cornerRadius(10)
                
            }

        }
        .padding()
        .frame(height: height)
        .background(ColorManager.BackgroundColor)
    }
}
